### PR TITLE
[osg] Cleanup, export unofficial config

### PR DESCRIPTION
**Pull request by dg0yt**
- #### What does your PR fix?
  Maintenance improvements.
   - Major change: `osg` and `osgearth` have custom find modules and an unusual approach to `<PKG>_LIBRARY_DEBUG`/`<PKG>_LIBRARY` when asked to link the variable name `<PKG>_LIBRARY`. This PR adds the preference to use `<PKG>_LIBRARIES` instead, which carries debug+release+dependencies in standard find modules (with help of vcpkg wrappers.) This change is the key to reduce patching. It is also no longer necessary to pass parameters about static dependencies, because they only activate uneducated guessing about transitive dependencies.
   - Revises features and dependencies for more control, leveraging `supports`.
   - Exports unofficial CMake config from `osg`, and uses it for `osgearth`. By using the targets, we don't need to patch the order of libraries, and we also get the transitive usage requirements.
   - Builds apps (`tools`) only for debug.
   - Collects plugins in `/plugins/osgPlugins-<VERSIONS>`.

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?
  no changes

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
  yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?
  yes

### CODE REVIEW

**Comment by dg0yt** [comment]
To be continued after merge of #24888 which fixes installation order and linkage issues with port nvtt.

**Comment by dg0yt** [comment]
Some libs in exported config which potentially need to replaced by targets to capture usage requirements:
~~~
"${_IMPORT_PREFIX}/debug/lib/libz.a"
"${_IMPORT_PREFIX}/debug/lib/libjpeg.a"
"${_IMPORT_PREFIX}/debug/lib/libjasperd.a"
"${_IMPORT_PREFIX}/debug/lib/libgif.a"
"${_IMPORT_PREFIX}/debug/lib/libpng16d.a"
"${_IMPORT_PREFIX}/debug/lib/libtiffd.a"
"${_IMPORT_PREFIX}/debug/lib/libgdal.a"
"${_IMPORT_PREFIX}/debug/lib/libgta.a"
"${_IMPORT_PREFIX}/debug/lib/libcurl-d.a"
"${_IMPORT_PREFIX}/debug/lib/static/libnvtt_d.a"
"${_IMPORT_PREFIX}/debug/lib/static/libnvcore_d.a"
"${_IMPORT_PREFIX}/debug/lib/static/libnvmath_d.a"
"${_IMPORT_PREFIX}/debug/lib/static/libnvimage_d.a"
"${_IMPORT_PREFIX}/debug/lib/libfreetyped.a"
"${_IMPORT_PREFIX}/debug/lib/libbz2d.a"
"${_IMPORT_PREFIX}/debug/lib/libbrotlidec-static.a"
"${_IMPORT_PREFIX}/debug/lib/libbrotlicommon-static.a"
~~~

**Comment by dg0yt** [comment]
I think this is a mergeable state, to unblock updates to gdal and osgearth.

File: ports/osgearth/portfile.cmake [hunk 1]
```diff
@@ -1,83 +1,93 @@
...
+if(VCPKG_LIBRARY_LINKAGE STREQUAL "static")
+    vcpkg_replace_string("${CURRENT_PACKAGES_DIR}/include/osgEarth/Export" "defined( OSGEARTH_LIBRARY_STATIC )" "1")
+endif()
 
-    file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/bin/${OSG_PLUGINS_SUBDIR}")
+# Merge osgearth plugins into [/debug]/plugins/osgPlugins-${OSG_VER},
+# as a staging area for later deployment.
+if(WIN32)
+    set(osg_plugin_pattern "osgdb*.dll")
+elseif(APPLE)
```
Comment (JackBoosY) [inline ports/osgearth/portfile.cmake:62]:
```suggestion
if(VCPKG_TARGET_IS_WINDOWS)
    set(osg_plugin_pattern "osgdb*.dll")
elseif(VCPKG_TARGET_IS_OSX)
```

File: ports/osg/portfile.cmake [hunk 2]
```diff
@@ -7,167 +7,160 @@ vcpkg_from_github(
...
-    file(COPY ${OSG_PLUGINS_DBG} DESTINATION ${CURRENT_PACKAGES_DIR}/debug/tools/${PORT}/osgPlugins-${OSG_VER})
-endif()
+if("tools" IN_LIST FEATURES)
+    if(WIN32)
+        set(osg_plugin_pattern "osgdb*.dll")
+    elseif(APPLE)
+        set(osg_plugin_pattern "libosgdb*.dylib")
+    else()
+        set(osg_plugin_pattern "libosgdb*.so")
+    endif()
```
Comment (JackBoosY) [inline ports/osg/portfile.cmake:139]:
```suggestion
    if(VCPKG_TARGET_IS_WINDOWS)
        set(osg_plugin_pattern "osgdb*.${VCPKG_TARGET_SHARED_LIBRARY_SUFFIX}")
    else()
        set(osg_plugin_pattern "libosgdb*.${VCPKG_TARGET_SHARED_LIBRARY_SUFFIX}")
    endif()
```

Comment (JackBoosY) [inline ports/osg/portfile.cmake:139]:
```suggestion
    if(VCPKG_TARGET_IS_WINDOWS)
        set(osg_plugin_pattern "osgdb*${VCPKG_TARGET_SHARED_LIBRARY_SUFFIX}")
    else()
        set(osg_plugin_pattern "libosgdb*${VCPKG_TARGET_SHARED_LIBRARY_SUFFIX}")
    endif()
```

File: ports/osg/unofficial-osg-config.cmake [hunk 3]
```diff
@@ -0,0 +1,20 @@
+cmake_policy(PUSH)
+cmake_policy(SET CMP0012 NEW)
+cmake_policy(SET CMP0054 NEW)
+cmake_policy(SET CMP0057 NEW)
+include("${CMAKE_CURRENT_LIST_DIR}/osg-targets.cmake")
+include(CMakeFindDependencyMacro)
+if("@OSG_TEXT_USE_FONTCONFIG@")
+    find_dependency(Fontconfig) # CMake 3.14
```
Comment (JackBoosY) [inline ports/osg/unofficial-osg-config.cmake:8]: Should we make sure users are using at least cmake 3.14 here?

File: ports/osgearth/portfile.cmake [hunk 4]
```diff
@@ -1,83 +1,93 @@
...
+    set(osg_plugin_pattern "libosgdb*.dylib")
+else()
+    set(osg_plugin_pattern "libosgdb*.so")
+endif()
```
Comment (JackBoosY) [inline ports/osgearth/portfile.cmake:66]:
```suggestion
if(VCPKG_TARGET_IS_WINDOWS)
    set(osg_plugin_pattern "osgdb*${VCPKG_TARGET_SHARED_LIBRARY_SUFFIX}")
else()
    set(osg_plugin_pattern "libosgdb*${VCPKG_TARGET_SHARED_LIBRARY_SUFFIX}")
endif()
```

**Review by JackBoosY — CHANGES_REQUESTED** [review]
Did you test the usage?

File: ports/osg/unofficial-osg-config.cmake (hunk 3 above)
Comment (dg0yt) [reply ports/osg/unofficial-osg-config.cmake:8]: Given the complexity of the `osg` software stack, I don't feel a strong desire to add complexity in the ports just for users with old versions of CMake. (And this comment is written on Ubuntu 18.04 with CMake 3.10.)

**Comment by dg0yt** [comment]
> Did you test the usage?

Usage of `osg`? It happens in `osgearth'. (There is also `osg-qt`.) Even with heavy patching is `osgearth`, it was far from perfect before these changes. I started this PR because the pending GDAL update is blocked by `osg`. I hope it is all okay now, but at least it is better base for drive-by fixes.

Usage of `osgearth`? It exports neither CMake config nor pkg-config. Pretty unusable, but this is not a regressions.

**Review by JackBoosY — APPROVED** [review]

File: ports/osgearth/find-package.patch [hunk 5]
```diff
@@ -31,23 +26,18 @@ index 7541c8d07..cc14d2160 100755
 -find_package(Protobuf)
 -find_package(WEBP)
 -find_package(Blend2D)
-+find_package(geos CONFIG REQUIRED)
-+set(GEOS_LIBRARY GEOS::geos GEOS::geos_c)
-+set(GEOS_FOUND 1)
-+find_package(unofficial-sqlite3 CONFIG REQUIRED)
-+set(SQLITE3_FOUND 1)
++find_package(GEOS CONFIG REQUIRED)
```
Comment (ras0219-msft) [inline ports/osgearth/find-package.patch:29]: Can we replace most of this patch with `CMAKE_REQUIRE_FIND_PACKAGE_<PackageName>`?
I know we'd still need the various `set(Foo_LIBRARY ...)` calls.

https://cmake.org/cmake/help/latest/variable/CMAKE_REQUIRE_FIND_PACKAGE_PackageName.html

**Review by ras0219-msft — COMMENTED** [review]
Some questions before merging. Otherwise LGTM, thanks again for the awesome cleanups!

File: ports/osgearth/find-package.patch (hunk 5 above)
Comment (dg0yt) [reply ports/osgearth/find-package.patch:29]: I even had this at some point in time, given the depth of the dependency tree, I wasn't sure if I want to enforce it for nested calls.
(The focus of my PR was osg. This is osgearth, which is waiting for a separate update.)

File: ports/osgearth/vcpkg.json [hunk 6]
```diff
@@ -31,9 +31,29 @@
     }
   ],
   "default-features": [
+    "default-features",
```
[the remark this answers is not in the document]
Comment (dg0yt) [reply ports/osgearth/vcpkg.json:34]: I didn't want to remove `tools` in a micro update. In addition, building executables is the ultimate test to check if transitive usage requirements are properly carried. So this feature is precious in vcpkg CI.
(The focus of my PR was osg. This is osgearth, which is waiting for a separate update.)

File: ports/osg/unofficial-osg-config.cmake [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+endif()
+if(${CMAKE_FIND_PACKAGE_NAME}_FIND_COMPONENTS AND "plugins" IN_LIST ${CMAKE_FIND_PACKAGE_NAME}_FIND_COMPONENTS)
```
[the remark this answers is not in the document]
Comment (dg0yt) [reply ports/osg/unofficial-osg-config.cmake:10]: Yes. Can't say why I added it.

File: ports/osgearth/find-package.patch (hunk 5 above)
Comment (dg0yt) [reply ports/osgearth/find-package.patch:29]: For example, there is:
~~~
# Mobile/GLES:
IF (OSGEARTH_USE_GLES)
    find_package(OpenGLES)
    find_package(OpenGLES REQUIRED)
ELSE ()
    find_package(OpenGL)
    find_package(OpenGL REQUIRED)
ENDIF (OSGEARTH_USE_GLES)
~~~
The packages are alternative, so IMO it is not a good idea to make both packages required at the same time, in the whole build system.

File: ports/osgearth/vcpkg.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (ras0219-msft) [reply ports/osgearth/vcpkg.json:34]: If the needed purpose is to build things in vcpkg ci, we should add a test port for it in `scripts/test_ports/` and not saddle every user with building things they don't need.

Please remove `"tools"` from the default features list; it's very easy for a consumer to add it back in as a dependency that's forward and backwards compatible.

**Review by ras0219-msft — CHANGES_REQUESTED** [review]

File: ports/osgearth/vcpkg.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (dg0yt) [reply ports/osgearth/vcpkg.json]: Done.

**Pull request merged**